### ISSUE TITLE
Issue with type conversion for BatchNorm inference in THNN

### ISSUE THREAD

**Issue by libfun** [pytorch/pytorch#3624]
TL;DR:
THNN/BatchNormalization.c, Line 53
https://github.com/pytorch/pytorch/blob/5478d0154fca6859c2bd6c8147b6797074d1efcf/aten/src/THNN/generic/BatchNormalization.c#L53

The variable `eps` is double, so `THTensor_(get1d)(running_var, f)` converts to double, so whole `1 / sqrt` thing becomes double, and in the end converts to `real`, because `invstd` is `real`.

If eps is converted to `float` before, the output should match the expected behavior.

I was trying to convert Pytorch model to Caffe on CPU and all conv layers worked flawlessly and produced equal output tensors, but on BatchNorm layers something broke and the output was very different. After some investigation I found that Caffe implementation produced the same output as direct numpy equivalent in float32, but Pytorch didn't. After setting types to float64 for values described above I was able to reproduce Pytorch batchnorm output with numpy.

This issue breaks direct conversion of Pytorch models to Caffe, and also seems like genuine unexpected behavior, since everything is expected be float32 during both training and inference.

**Comment by libfun** [comment]
Xposted issue in torch/nn: https://github.com/torch/nn/issues/1296

**Comment by colesbury** [comment]
Is the issue is that the intermediate value is computed with 64-bits of precision instead of 32-bits? Why is that a problem?

**Comment by SsnL** [comment]
How different is "very different"?

**Comment by libfun** [comment]
"Very different" comment was before I found other issue with the order of elemwise sum and nonlinearity in caffe conversion. Without it overall the highest per-pixel error was on the order of 1e-7, so it's not significant at all.
I don't believe this issue to be very important, just that it behaves not like you would expect, when you're working with float32 precision.

**Comment by SsnL** [comment]
Good to hear. Unfortunately moving models across frameworks can have such issues. But usually the different is rather small as you observed.

**Comment by ngimel** [comment]
On the cpu, accreal type for float32 inputs is double, so in many cases for fp32 inputs intermediate computations will be in double. It indeed may be unexpected and not match numpy behavior, but the differences should be small in most cases.

**Comment by libfun** [comment]
> On the cpu, accreal type for float32 inputs is double, so in many cases for fp32 inputs intermediate computations will be in double
Shouldn't that mean that the outputs computed with cunn and on cpu should differ?

**Comment by colesbury** [comment]
I'm not sure what "matches NumPy behavior" means here. There's no NumPy implementation of BatchNorm and even if there were, we would not match it exactly. Different implementations will produce slightly different results. You'll get different results for different ordering of additions not to mention [x87 vs SSE](http://christian-seiler.de/projekte/fpmath/) fpmath . 

We never guarantee that float32 functions only use float32 computations. In many places we use double-precision floats on the CPU for reductions before converting back to single-precision.

**Comment by colesbury** [comment]
@libfun, yes they differ: our CUDA functions do not produce the exact same results ad the CPU functions.
